packages/ethereum/contracts/src/node-stake/NodeStakeFactory.sol

```diff
@@ -0,0 +1,114 @@
+// SPDX-License-Identifier: GPL-3.0
+pragma solidity ^0.8.0;
+
+import "openzeppelin-contracts-upgradeable/proxy/ClonesUpgradeable.sol";
+import "openzeppelin-contracts/utils/Address.sol";
+import "../../script/utils/SafeSuiteLib.sol";
+import "safe-contracts/proxies/SafeProxy.sol";
+import "safe-contracts/proxies/SafeProxyFactory.sol";
+import "safe-contracts/Safe.sol";
+import "safe-contracts/common/Enum.sol";
+
+
+contract HoprNodeStakeFactory  {
+    using Address for address;
+    using ClonesUpgradeable for address;
+
+    address internal constant SENTINEL_OWNERS = address(0x1);
+    bytes32 internal r;
+    bytes internal approvalHashSig;
+
+    error TooFewOwners();
+    event NewHoprNodeStakeModule(address instance);
+    event NewHoprNodeStakeSafe(address instance);
+
+    constructor() {
+        r = bytes32(uint256(uint160(address(this))));
+        approvalHashSig = abi.encodePacked(abi.encode(r, bytes32(0)), bytes1(hex"01"));
+    }
+
+    /**
+     * @dev Returns the version of Safe deployments
+     */
+    function safeVersion() public pure returns (string memory) {
+        return SafeSuiteLib.SAFE_VERSION;
+    }
+
+    /**
+     * @dev Create a safe proxy and a module proxy
+     * @param moduleSingletonAddress singleton contract of Safe
+     * @param admins owners of safe by default it's 1 out of n
+     * @param nonce nonce to create salt
```

Comment:
if we let users give customize nonce, shouldn't we let them give two nonces for both the node module and the Safe deployment?